### PR TITLE
Add Bessel functions to math library

### PR DESCRIPTION
**Pull request by BeyondEspresso**
This pull request adds Bessel-functions from the `CUDA` and `BOOST` libraries to the `PMACC` math implementation. Up to now, there have been no extensive tests, but it seems to compile fine so far.

This contribution does not have any special priority. So, please have a look at the code and give feedback at your leisure. :smiley:

### CODE REVIEW

File: include/pmacc/algorithms/math/doubleMath/bessel.tpp [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+{
+namespace algorithms
+{
+namespace math
+{
+
+    template< >
+    struct Cyl_bessel_i0< double >
+    {
+    typedef double result;
```
Comment (psychocoderHPC) [inline include/pmacc/algorithms/math/doubleMath/bessel.tpp]: please indent all class/Struct member methods and variables.

File: include/pmacc/algorithms/math/defines/bessel.hpp [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    {
+        return Cyl_bessel_i0< T_Type >( )( x );
+    }
+
+    /* Bessel function of first kind of order 0 */
+
+    template< typename T_Type >
+    struct J0;
+
+    /**
```
Comment (psychocoderHPC) [inline include/pmacc/algorithms/math/defines/bessel.hpp]: must be 

```
/** Calculate the value of the Bessel function
 *
 * of first kind of order 0 for the input argument.
 *
 * @return float value
 */
```

Comment (psychocoderHPC) [inline include/pmacc/algorithms/math/defines/bessel.hpp]: please add also a documentation or `x` with `@param x ...`

File: include/pmacc/algorithms/math/defines/bessel.hpp [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+     */
+    template< typename T_Type >
+    HDINLINE typename J0< T_Type >::result j0( T_Type x )
+    {
+        return J0< T_Type >( )( x );
+    }
+
+    /* Bessel function of first kind of order 1 */
+
+    template< typename T_Type >
```
Comment (psychocoderHPC) [inline include/pmacc/algorithms/math/defines/bessel.hpp]: Please add a documentation to the forward type.

**Comment by psychocoderHPC** [comment]
There should no struct or function called `J1` or `j1` within the `math` namespace. Please add a namepsace `bessel` or give it a more characterizing name.

File: include/pmacc/algorithms/math/defines/bessel.hpp [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+namespace math
+{
+
+    /* Modified cylindrical Bessel function of order 1 */
+
+    template< typename T_Type >
+    struct Cyl_bessel_i1;
+
+    /**
+     * Calculate the value of the regular modified cylindrical Bessel function
```
Comment (ax3l) [inline include/pmacc/algorithms/math/defines/bessel.hpp]: here and below I don't understand what the adjective "regular" shall stand for

File: include/pmacc/algorithms/math/defines/bessel.hpp [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+
+namespace pmacc
+{
+namespace algorithms
+{
+namespace math
+{
+
+    /* Modified cylindrical Bessel function of order 1 */
+
```
Comment (ax3l) [inline include/pmacc/algorithms/math/defines/bessel.hpp]: remove this newline & make it a normal doxygen comment?

File: include/pmacc/algorithms/math/defines/bessel.hpp [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+     * @return float value
+     */
+    template<typename T_Type >
+    HDINLINE typename Cyl_bessel_i1<T_Type>::result cyl_bessel_i1( T_Type x )
+    {
+        return Cyl_bessel_i1< T_Type >( )( x );
+    }
+
+    /* Modified cylindrical Bessel function of order 0 */
+
```
Comment (ax3l) [inline include/pmacc/algorithms/math/defines/bessel.hpp]: remove this newline & make it a normal doxygen comment?

File: include/pmacc/algorithms/math/defines/bessel.hpp [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+     * @return float value
+     */
+    template< typename T_Type >
+    HDINLINE typename Cyl_bessel_i0< T_Type >::result cyl_bessel_i0( T_Type x )
+    {
+        return Cyl_bessel_i0< T_Type >( )( x );
+    }
+
+    /* Bessel function of first kind of order 0 */
+
```
Comment (ax3l) [inline include/pmacc/algorithms/math/defines/bessel.hpp]: remove this newline & make it a normal doxygen comment?

File: include/pmacc/algorithms/math/defines/bessel.hpp [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+     * @return float value
+     */
+    template< typename T_Type >
+    HDINLINE typename J0< T_Type >::result j0( T_Type x )
+    {
+        return J0< T_Type >( )( x );
+    }
+
+    /* Bessel function of first kind of order 1 */
+
```
Comment (ax3l) [inline include/pmacc/algorithms/math/defines/bessel.hpp]: remove this newline & make it a normal doxygen comment?

File: include/pmacc/algorithms/math/defines/bessel.hpp [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+    template< typename T_Type >
+    struct Cyl_bessel_i1;
+
+    /**
+     * Calculate the value of the regular modified cylindrical Bessel function
+     * of order 1 for the input argument.
+     * @return float value
+     */
+    template<typename T_Type >
+    HDINLINE typename Cyl_bessel_i1<T_Type>::result cyl_bessel_i1( T_Type x )
```
Comment (ax3l) [inline include/pmacc/algorithms/math/defines/bessel.hpp]: all the `x` arguments in these wrappers need to be const references

File: include/pmacc/algorithms/math/floatMath/bessel.tpp [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+{
+namespace algorithms
+{
+namespace math
+{
+
+    template< >
+    struct Cyl_bessel_i0< float >
+    {
+    typedef float result;
```
Comment (ax3l) [inline include/pmacc/algorithms/math/floatMath/bessel.tpp]: C++11 `using` please ;)

File: include/pmacc/algorithms/math/doubleMath/bessel.tpp (hunk 1 above)
Comment (ax3l) [inline include/pmacc/algorithms/math/doubleMath/bessel.tpp]: C++11 `using` please :)

**Comment by ax3l** [comment]
@BeyondEspresso a comment from @psychocoderHPC and me: it would make sense to group the new functions in a namespace `bessel` and maybe even in `cylindrical` so that functions like `J0` do not go un-prefixed in the `math` namespace.

File: include/pmacc/algorithms/math/defines/bessel.hpp (hunk 4 above)
Comment (BeyondEspresso) [inline include/pmacc/algorithms/math/defines/bessel.hpp]: It means the same thing as *first kind*. I'll rewrite it accordingly.

**Comment by BeyondEspresso** [comment]
Thank you for the first feedback. I included the review suggestions from @ax3l and @psychocoderHPC .

**Comment by psychocoderHPC** [comment]
Please add a define to the generic bessel header with the information if the bessel function is available. The reason for that is that the bessel functions is introduced with CUDA8 and not supported in CUDA 7.5. ([source](https://developer.nvidia.com/cuda-math-library))

```
#if !defined(__CUDA_ARCH__) && ( __CUDACC_VER_MAJOR__ >= 8 )
#   define PMACC_HAS_MATH_BESSEL 1
#else
#   definePMACC_HAS_MATH_BESSEL 0
#endif
```

**Comment by BeyondEspresso** [comment]
The bessel functions called by this PR have been partially implemented since CUDA 4.1 and are complete by CUDA 6.0 . Do I still need to add a define?

**Comment by psychocoderHPC** [comment]
@BeyondEspresso  You are right the bessel functions are available in CUDA 7.5.

File: include/pmacc/algorithms/math/floatMath/bessel.tpp [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+            return boost::math::cyl_bessel_i(
+                1,
+                x
+            );
+#endif
+        }
+    };
+
+    template< >
+    struct J0<float>
```
Comment (psychocoderHPC) [inline include/pmacc/algorithms/math/floatMath/bessel.tpp]: missing space after `<` and before`>`

File: include/pmacc/algorithms/math/floatMath/bessel.tpp [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+            return boost::math::cyl_bessel_j(
+                0,
+                x
+            );
+#endif
+        }
+    };
+
+    template< >
+    struct J1<float>
```
Comment (psychocoderHPC) [inline include/pmacc/algorithms/math/floatMath/bessel.tpp]: missing space after `<` and before`>`

File: include/pmacc/algorithms/math/floatMath/bessel.tpp [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+            return boost::math::cyl_bessel_j(
+                1,
+                x
+            );
+#endif
+        }
+    };
+
+    template< >
+    struct Jn<int, float>
```
Comment (psychocoderHPC) [inline include/pmacc/algorithms/math/floatMath/bessel.tpp]: - missing space after `<` and before`>`
- please use one line per argument

File: include/pmacc/algorithms/math/floatMath/bessel.tpp [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+            return boost::math::cyl_bessel_j(
+                n,
+                x
+            );
+#endif
+        }
+    };
+
+    template< >
+    struct Y0<float>
```
Comment (psychocoderHPC) [inline include/pmacc/algorithms/math/floatMath/bessel.tpp]: missing space after `<` and before`>`

File: include/pmacc/algorithms/math/floatMath/bessel.tpp [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+            return boost::math::cyl_neumann(
+                1,
+                x
+            );
+#endif
+        }
+    };
+
+    template< >
+    struct Yn< int, float >
```
Comment (psychocoderHPC) [inline include/pmacc/algorithms/math/floatMath/bessel.tpp]: - missing space after `<` and before`>`
- please use one line per argument

**Comment by BeyondEspresso** [comment]
Done. :smiley:

**Pull request merged**